### DISCUSSION TITLE
prob cat thresh in point2grid

### DISCUSSION THREAD

**Discussion by cjmelick** [dtcenter/METplus#1845] [Configuration] [answered]
Greetings!  I'm having problems setting a threshold to my input point obs data when using prob cat thresh in point2grid.   I have the method set to GAUSSIAN to create a probabilistic analysis.   The only thing I could get to work was setting the value to NA or not setting anything explicitly (which resulted in what looks like an NA implicit threshold).   What are the possible entry options?  My input observations are binary, either zero or one.  Please let me know if you need more information.   

Thanks,
Chris

**Comment by JohnHalleyGotway** [comment, marked as answer]
Hi Chris, I see you have a question about the point2grid tool. I took at look at one of the unit tests we run to test out this tool that are defined in unit_point2grid.xml. This one named "[point2grid_pb2nc_TMP_prob](https://github.com/dtcenter/MET/blob/main_v10.1/test/xml/unit_point2grid.xml#L64)" sounds pretty similar to what you describe.

I ran the following command to test it out:
```
point2grid pb2nc/ndas.20120409.t12z.prepbufr.tm00.nc G212 point2grid/pb2nc_TMP_prob_gaussian.nc \
-field 'name="TMP"; level="*"; valid_time="20120409_12";' -name TEMP -prob_cat_thresh ">=280.0" -method GAUSSIAN
```
And then plotted this field from the output file:
```
plot_data_plane point2grid/pb2nc_TMP_prob_gaussian.nc TEMP_prob_ge280.ps 'name="TEMP_prob_ge280.0";level="(*,*)";'
```
And here's the resulting image:
![Screen Shot 2022-10-03 at 4 58 15 PM](https://user-images.githubusercontent.com/21087144/193700458-6b527c0c-e7cd-414a-b8a9-36384e91fdbf.png)

For your binary input data, I'd recommend defining the threshold as `-prob_cat_thresh eq1`.
Does that help you make it any further?

**Reply by cjmelick** [reply]
Hi John,

I think that does help -- I will have to get back to you after I try some tests.   I never did place a logical operator within the argument, which makes perfect sense now!   What's the difference between the GAUSSIAN and GAUSSMAX methods and what/how exactly is the original gridded field that is produced, such as in the temperature field above (not the probabilistic field)?

Thanks,
Chris

**Reply by JohnHalleyGotway** [reply]
Here's the result of re-running with GAUSSMAX instead of GAUSSIAN:
```
point2grid pb2nc/ndas.20120409.t12z.prepbufr.tm00.nc G212 \
point2grid/pb2nc_TMP_prob_gaussmax.nc \
-field 'name="TMP"; level="*"; valid_time="20120409_12";' \
-name TEMP -prob_cat_thresh ">=280.0" -method GAUSSMAX
plot_data_plane point2grid/pb2nc_TMP_prob_gaussmax.nc \
point2grid/pb2nc_TMP_prob_gaussmax.ps \
'name="TEMP_prob_ge280.0"; level="(*,*)";'
```
<img width="438" alt="Screen Shot 2022-10-04 at 11 16 59 AM" src="https://user-images.githubusercontent.com/21087144/193884198-ee089b12-04fb-40ea-84c7-1574f9700399.png">

Instead of having values between 0 and 1, we just get a bunch of 0's and 1's.
Unfortunately, I'm not the right person to comment on or interpret that result.

However, I believe that GAUSSIAN is only supported interpolating points, not regridding. While MAXGAUSS is appropriate for regridding. However, I could ask one of the scientists on the team to provide a better explanation.

**Reply by cjmelick** [reply]
John,

Thanks for the demonstration and explanation for the alternative method option!  If you could get further details, I would appreciate it.   Two other questions: 1.) in your specific case, what does the temperature field itself look like when you perform the GAUSSIAN versus GAUSSMAX option? 2.) Is it possible to do multiple categorical thresholds from one system call to point2grid?

Chris

**Reply by JohnHalleyGotway** [reply]
@TaraJensen and @CPKalb, who would be the best scientist to advise on the configuration and interpretation of the GAUSSIAN and MAXGAUSS methods for the point2grid tool?

The point2grid [usage statement](https://met.readthedocs.io/en/latest/Users_Guide/reformat_point.html?highlight=prob_cat_thresh#optional-arguments-for-point2grid) does not indicate that `-prob_cat_thresh` can be used multiple times. So it's limited to processing a single threshold and generating a single output "probability" field. I can certainly see how the tool could be enhanced to process multiple thresholds, but that's not what it does currently. That'd require an enhancement.

So for now, you're stuck running point2grid once for each threshold you'd like to process.

In the example I sent, the temperature data is just point observations from many stations. I don't actually think we're even filtering them by message type or vertical level in this software test. But to demonstrate what the input point data looks like, I ran `plot_point_obs` like this:
```
plot_point_obs pb2nc/ndas.20120409.t12z.prepbufr.tm00.nc point_tmp.ps \
-plot_grid G212 -config PlotPointObsConfig_TMP
```
Using [PlotPointObsConfig_TMP](https://github.com/dtcenter/METplus/files/9710608/PlotPointObsConfig_TMP.txt)

And here's the resulting image:
<img width="609" alt="Screen Shot 2022-10-04 at 2 47 30 PM" src="https://user-images.githubusercontent.com/21087144/193925139-d1de2ed9-25c5-460c-93c3-9b041075eade.png">

For real scientific applications, you really shouldn't mix surface and upper-air temps together!

**Reply by cjmelick** [reply]
John, that's fine about the limitations. It would be a great enhancement, though, I could see.   

I meant what does the gridded temperature field look like and how exactly is that constructed when using GAUSSIAN or GAUSSMAX?  I guess I could see that for GAUSSMAX it might just be the MAX value for the observations that fall within the vicinity of the grid point.   

Thanks for your help,
Chris

**Reply by cjmelick** [reply]
Hi John,

I never heard back from anyone about the technical approach for GAUSSIAN and MAXGAUSS and how they differ.  I noticed that you still get the variable field requested as well before I guess the gaussian kernel is applied.   How is the variable field different than the probabilistic field when the threshold is applied?   How is the original field determined?  I guess in terms of the MAXGAUSS it might just be the maximum value before the gaussian kernel is applied.   Also, I noticed that the vld_thresh is available in point2grid?  How is this argument utilized within the point observations when deciding about what the final grid point value will be?   

I hope I made the questions clear.   Thanks for all of your help!  
Chris

**Reply by JohnHalleyGotway** [reply]
1. How doe the GAUSSIAN and MAXGAUSS methods differ?

Example of running point2grid on point observations of TMP with `-prob_cat_thresh >=280`:
```
point2grid pb2nc/ndas.20120409.t12z.prepbufr.tm00.nc G212 point2grid/pb2nc_TMP_prob_gaussian.nc \
-field 'name="TMP"; level="*"; valid_time="20120409_120000";' \
-name TEMP -prob_cat_thresh ">=280.0" -method GAUSSIAN
```
And with MAXGAUSS:
```
point2grid pb2nc/ndas.20120409.t12z.prepbufr.tm00.nc G212 point2grid/pb2nc_TMP_prob_maxgauss.nc \
-field 'name="TMP"; level="*"; valid_time="20120409_120000";' \
-name TEMP -prob_cat_thresh ">=280.0" -method MAXGAUSS
```
Both NetCDF files contain the a `float TEMP(lat, lon) ;` variable but their values differ for GAUSSIAN (left) and MAXGAUSS (right) as seen in these plots.
<img width="1129" alt="Screen Shot 2022-11-17 at 10 19 00 AM" src="https://user-images.githubusercontent.com/21087144/202513896-4933d919-f26e-4400-b81f-0423cb905f50.png">

The reason can be found in the log messages:
GAUSSIAN: `DEBUG 2: Interpolation options: method = UW_MEAN, vld_thresh = 0.5`
MAXGAUSS: `DEBUG 2: Interpolation options: method = MAX, vld_thresh = 0.5`

See this [line of code](https://github.com/dtcenter/MET/blob/61e192883f76bf7362152c19c2d279fa9d5dc5c7/met/src/tools/other/point2grid/point2grid.cc#L2854).

When multiple obs fall in the same grid both, GAUSSIAN stores their unweighted mean value while MAXGAUSS stores the maximum value.

The downstream logic is the same. If provided, apply the `-prob_cat_thresh` threshold to convert the derived field to a binary field of 0's and 1's. Then apply the gaussian smoother on the result to generate "probability" fields:

<img width="1124" alt="Screen Shot 2022-11-17 at 10 23 42 AM" src="https://user-images.githubusercontent.com/21087144/202514919-4d278684-30a8-4ef0-bc0e-24de1647a9d1.png">

So GAUSSIAN vs MAXGAUSS yields one small difference in the logic... computing the mean of the values in each grid box or the maximum.

2.  Also, I noticed that the vld_thresh is available in point2grid? How is this argument utilized within the point observations when deciding about what the final grid point value will be?

The `vld_thresh` option is used in this call to the [interp_gaussian_dp()](https://github.com/dtcenter/MET/blob/61e192883f76bf7362152c19c2d279fa9d5dc5c7/met/src/tools/other/point2grid/point2grid.cc#L1053) function.

So it *SHOULD* impact how much the edge effects impact the values inside the domain. However, testing with `-vld_thresh 0` and `-vld_thresh 1`, I see that it has no impact on the output. And I see why.

The processing steps are:

1. Create a grid consisting of all bad data using the user-specified grid.
2. Loop through the point observations and "store" them in the grid cell in which they fall.
3. Summarize the observation values inside each grid cell (e.g. compute the maximum or mean value). Write the result to the `TEMP` variable in my example which contains bad data values surrounded by valid data values.
4. Threshold the TEMP field to create a 0/1 mask field. But here's the issue... *thresholding the bad data values results in a mask value of 0 rather than bad data.*
5. Apply the gaussian filter to that mask field which is entirely 0's and 1's, with no bad data.

So even though `-vld_thresh` is configurable, no bad data values survive the masking step. The gaussian smoother doesn't have any bad data values to decided how to handle. So changing `-vld_thresh` has no impact on the results.

So the question is really... when we apply the `-prob_cat_thresh` what should we do with grid points in which no point observations fell? Masking with a value of 0 (current behavior) makes the `-vld_thesh` argument pointless. Passing the bad data value through the mask will have a LARGE impact on the results.

I think this highlights the issues introduced by placing relatively sparse point data onto a relatively dense grid.

**Reply by cjmelick** [reply]
John, 

Thank you for the excellent description and examples!   So, is -vld_thresh only useful when calling the GUASSIAN or MAXGAUSS method (if there is bad data)?   And when you say bad data points, is that where the undefined missing value comes into play?

Chris

**Reply by JohnHalleyGotway** [reply]
Let me illustrate with some more pictures. The following NetCDF variables show up in the output. In this case, it's with `-method MAXGAUSS`:
```
	int TEMP_cnt(lat, lon) ;
	float TEMP(lat, lon) ;
	int TEMP_mask(lat, lon) ;
	float TEMP_prob_ge280.0(lat, lon) ;
```

1. The `int TEMP_cnt(lat, lon) ;` variable contains counts of observations falling inside each grid box. They are integers ranging from 0 to 111.
<img width="521" alt="Screen Shot 2022-11-17 at 11 39 35 AM" src="https://user-images.githubusercontent.com/21087144/202530653-a3867268-9a51-45d1-9926-3f5dc132b74e.png">

2. The `int TEMP_mask(lat, lon) ;` variable is just 0/1 indicating where TEMP_cnt > 0.
<img width="492" alt="Screen Shot 2022-11-17 at 11 42 32 AM" src="https://user-images.githubusercontent.com/21087144/202531180-ad778ac7-25bb-4a5d-8826-1128301ceea0.png">

3. The `float TEMP(lat,lon) ;` variable contains the maximum (because of MAXGAUSS) of the values falling inside each grid box. Grid cells where TEMP_cnt == 0 have a value of bad data which is plotted as dark gray, and stored internally in MET with a value of -9999.
<img width="566" alt="Screen Shot 2022-11-17 at 11 41 01 AM" src="https://user-images.githubusercontent.com/21087144/202530901-6e7e68be-1ed7-4474-9c37-2e97a14ddfd7.png">

4. The `TEMP_prob_ge280.0(lat,lon) ;` variable is the result of thresholding the TEMP >= 280 and then applying the gaussian smoother.
<img width="564" alt="Screen Shot 2022-11-17 at 11 48 13 AM" src="https://user-images.githubusercontent.com/21087144/202532248-9863530e-d31c-4c1d-9c29-e442bd923eb7.png">

As far as I can tell, the `-vld_thresh` option would only apply in point2grid in step 4. BUT it actually has no effect there because the field that is smoothed will never contain any bad data values. It will only ever contain 0's and 1's, not bad data.

I think we should reconsider these details very carefully and decide where we should remove the `-vld_thresh` option entirely or revise the processing logic in some way.

**Reply by cjmelick** [reply]
Hi John, that does illustrate the point.   I was sort of curious of whether that argument was even really being used in point2grid.

Thank you again,
Chris

**Reply by JohnHalleyGotway** [reply]
Sure thing, and sorry for the long delay in getting this answer to you. We've had a pretty hectic development cycle lately!

Please feel free to find/select the "mark as answer" button, and then I'll lock the discussion to close it out.

**Reply by cjmelick** [reply]
No worries and will do!
